Add MCP server support to kubectl-datadog plugin (#2456)
DataDog/datadog-operator: docs/kubectl-datadog-mcp.md

## Overview

The `kubectl datadog mcp` command starts an MCP (Model Context Protocol) server that provides read-only access to DatadogAgent resources in your Kubernetes cluster. The server integrates with AI assistants like Claude Desktop, allowing them to query and understand your Datadog Agent deployments.

The MCP server provides two types of tools:
- **Local tools**: Query DatadogAgent resources from the Kubernetes API
- **Proxy tools**: Automatically discovered and forwarded from the cluster-agent's MCP server for runtime diagnostics

## What is MCP?

The Model Context Protocol (MCP) is an open standard that enables AI assistants to securely connect to external data sources and tools. The `kubectl-datadog` MCP server exposes DatadogAgent resources through a standardized interface that AI assistants can query.

## Features

- **Read-only access**: All operations are safe GET/LIST operations
- **Namespace-aware**: Respects your kubeconfig context and namespace settings
- **Auto-selection**: Most tools automatically select a DatadogAgent when name is omitted (convenient for single-agent clusters)
- **Structured output**: Returns JSON data optimized for AI assistant consumption

> - **Local tools**: 5 specialized tools for querying DatadogAgent resources

Reviewer: I'd omit numbering, in case we add/remove tools in the future and would have to remember to update this line:
```suggestion
- **Local tools**: Specialized tools for querying DatadogAgent resources
```

- **Cluster-agent proxy**: Automatically discovers and exposes cluster-agent MCP tools for runtime diagnostics
- **Graceful degradation**: Local tools continue to work even if cluster-agent proxy is unavailable

## Installation

The MCP server is built into the `kubectl-datadog` plugin. No additional installation is required beyond having the kubectl-datadog plugin available in your PATH.

### Prerequisites

**Required**:
- `kubectl-datadog` plugin installed
- Valid kubeconfig with access to DatadogAgent resources
- Kubernetes cluster with Datadog Operator installed

**Optional (for cluster-agent proxy)**:
- Cluster-agent deployed and running
- Cluster-agent MCP server enabled (`cluster_agent.mcp.enabled=true`)
- RBAC permissions for `pods/list`, `pods/portforward`, and `leases/get` or `configmaps/get`

## Usage

### Starting the MCP Server

> To start the MCP server, simply run:

Reviewer: ```suggestion
To start the MCP server:
```

```bash
kubectl datadog mcp --proxy-port=5000
```

**Customize the cluster-agent MCP endpoint**:
```bash
kubectl datadog mcp --proxy-endpoint=/mcp
```

**Available flags**:
- `--proxy-cluster-agent`: Enable/disable proxy (default: true)
- `--proxy-dda-name`: DatadogAgent name to proxy (default: auto-select first)
- `--proxy-port`: Cluster-agent MCP server port (default: 5000)
- `--proxy-endpoint`: Cluster-agent MCP endpoint path (default: /mcp)

### Debug Logging

To enable debug logging for troubleshooting:

```bash
kubectl datadog mcp --debug
```

Debug logs are written to stderr and include:
- MCP server lifecycle events
- Proxy initialization steps
- Tool discovery and registration
- Port-forward connection status

In Claude Desktop configuration:
```json
{
  "mcpServers": {
    "kubectl-datadog": {
      "command": "kubectl",
      "args": ["datadog", "mcp", "--debug"]
    }
  }
}
```

## Configuration with Claude Desktop

To use the MCP server with Claude Desktop, add the following to your Claude Desktop configuration file:

**macOS/Linux**: `~/Library/Application Support/Claude/claude_desktop_config.json`

> **Windows**: `%APPDATA%\Claude\claude_desktop_config.json`

Reviewer: ```suggestion
To use the MCP server with Claude Desktop, add the following to your Claude Desktop configuration file. After updating the configuration, restart Claude Desktop for the changes to take effect

- **macOS/Linux**: `~/Library/Application Support/Claude/claude_desktop_config.json`
- **Windows**: `%APPDATA%\Claude\claude_desktop_config.json`
```

    "kubectl-datadog": {
      "command": "kubectl",
      "args": ["datadog", "mcp", "--proxy-dda-name", "my-datadog-agent"]
    }
  }
}
```

After updating the configuration, restart Claude Desktop for the changes to take effect.

## Available Tools

### Local Tools

The MCP server exposes 5 local tools that query DatadogAgent resources from the Kubernetes API.

**Auto-selection**: Tools that query a specific DatadogAgent (tools 2-5) support auto-selection. When the `name` parameter is omitted, the tool automatically selects the first DatadogAgent found in the namespace. This is convenient for clusters with a single DatadogAgent. If multiple agents exist, the first one is chosen and a warning is included in the error message.

### 1. list_datadog_agents

**Purpose**: List all DatadogAgent resources in a namespace or across all namespaces.

**Parameters**:
- `namespace` (optional): Kubernetes namespace to list agents from. Empty means current namespace.
- `allNamespaces` (optional): If true, list agents from all namespaces.

**Returns**: Summary of each agent including:
- Name and namespace
- Agent DaemonSet status
- Cluster Agent status
- Cluster Checks Runner status
- Age

> **Example prompt for Claude**:

Reviewer: In case we start supporting other assistants in addition to Claude (revise the lead-in line in other tools too):
```suggestion
**Example prompt**:
```

> "List all DatadogAgent resources in my cluster"

### 2. get_datadog_agent_status

**Purpose**: Get runtime status information for a DatadogAgent deployment.

**Parameters**:
- `name` (optional): Name of the DatadogAgent resource. If not specified, auto-selects the first DatadogAgent found in the namespace.
- `namespace` (optional): Kubernetes namespace. Empty means current namespace.

**Returns**: Detailed status for all components:
- Agent DaemonSet: desired, current, ready, available, up-to-date counts
- Cluster Agent: replica counts, deployment name
- Cluster Checks Runner: replica counts, deployment name
- Conditions: reconciliation status

> **Auto-selection**: When `name` is omitted, the tool automatically selects the first DatadogAgent in the namespace. If multiple agents exist, the first one is chosen and a warning is included.

Reviewer: Omit as repetitive? The `name` parameter above already explains how auto-selection works, and there's also a general explanation in the section overview above.

Looks like other tools follow a similar pattern, so consider omitting there too.

**Example prompts for Claude**:
> "What's the status of my datadog-agent deployment?"
> "Show me the DatadogAgent status" (uses auto-selection)

### 3. describe_datadog_agent_features

**Purpose**: View enabled features and their configuration.

**Parameters**:
- `name` (optional): Name of the DatadogAgent resource. If not specified, auto-selects the first DatadogAgent found in the namespace.
- `namespace` (optional): Kubernetes namespace. Empty means current namespace.

**Returns**: Feature configuration including:
- APM settings
- Log collection settings
- Network Performance Monitoring (NPM)
- Security features (CSPM, CWS, etc.)
- Process monitoring
- And 40+ other monitoring features

**Auto-selection**: When `name` is omitted, the tool automatically selects the first DatadogAgent in the namespace. If multiple agents exist, the first one is chosen and a warning is included.

**Example prompts for Claude**:
> "What monitoring features are enabled on datadog-agent?"

> > "Show me the enabled features" (uses auto-selection)

Reviewer: ```suggestion
> "What monitoring features are enabled on datadog-agent?"  
> "Show me the enabled features" (uses auto-selection)
```

- **Local only**: The stdio transport means the server only accepts connections from processes on the same machine (like Claude Desktop).
- **No authentication**: The server does not implement additional authentication beyond what kubectl provides.
- **Port-forward security**: The cluster-agent proxy uses Kubernetes port-forward, which creates a secure tunnel authenticated by your kubeconfig. The connection stays within your local machine and does not expose cluster-agent ports externally.
- **Proxy permissions**: The proxy requires additional RBAC permissions:
  - `pods/list`: To discover cluster-agent pods
  - `pods/portforward`: To establish port-forward connections
  - `leases/get` or `configmaps/get`: For leader election detection
- **Tool isolation**: Cluster-agent tools are prefixed with `cluster_agent_` to prevent naming conflicts with local tools.

## Troubleshooting

### MCP server not appearing in Claude Desktop

1. Verify your `claude_desktop_config.json` syntax is valid JSON
2. Ensure the `kubectl` command is in your PATH
3. Restart Claude Desktop after configuration changes

> 4. Check Claude Desktop logs for error messages

Reviewer: ```suggestion
1. Verify your AI assistant configuration file syntax is valid JSON
2. Ensure the `kubectl` command is in your PATH
3. Restart AI assistant after configuration changes
4. Check AI assistant logs for error messages
```

### "Failed to list DatadogAgents" errors

1. Verify your kubeconfig is valid: `kubectl config view`
2. Check you have permissions: `kubectl auth can-i list datadogagents`
3. Verify the Datadog Operator CRDs are installed: `kubectl get crd datadogagents.datadoghq.com`
4. Check the namespace exists and you have access

### Server disconnects immediately

1. Ensure kubectl-datadog is properly installed: `kubectl datadog --help`
2. Try running the command manually: `kubectl datadog mcp` (press Ctrl+C to exit)
3. Check for Go module errors or missing dependencies

### Claude doesn't seem to see my DatadogAgents

1. Verify agents exist: `kubectl get datadogagents -A`

> 2. Check the namespace Claude is querying (it uses your kubeconfig context)

Reviewer: ```suggestion
### AI assistant doesn't seem to see my DatadogAgents

1. Verify agents exist: `kubectl get datadogagents -A`
2. Check the namespace the AI assistant is querying (it uses your kubeconfig context)
```
